### PR TITLE
Vs add niches

### PR DESCRIPTION
**Pull request by Vadser**

### CODE REVIEW

File: app/controllers/niches_controller.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    @niche = current_user.niches.create
+  end
+
+  # GET /locations/1/edit
+  def edit; end
+
+  # POST /locations
+  # POST /locations.json
+  def create
+    parent = current_user.niches.find_by(id: niche_params[:parent])
```
Comment (Sergyenko) [inline app/controllers/niches_controller.rb]: What if current_user.niches returns 'nil'?

File: app/controllers/niches_controller.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    @niche = current_user.niches.create(name: niche_params[:name], parent: parent)
```
Comment (Sergyenko) [inline app/controllers/niches_controller.rb]: DRY 'current_user.niches'

File: app/controllers/niches_controller.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    if @niche.save
+      redirect_to niche_url(@niche)
```
Comment (Sergyenko) [inline app/controllers/niches_controller.rb]: would that be more elegant 'redirect_to @niche'

File: app/controllers/niches_controller.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    else
+      render :new
+    end
+  end
+
+  # PATCH/PUT /locations/1
+  # PATCH/PUT /locations/1.json
+  def update
+    parent = current_user.niches.find_by(id: niche_params[:parent])
```
Comment (Sergyenko) [inline app/controllers/niches_controller.rb]: DRY current_user.niches.find_by

File: app/controllers/niches_controller.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    if @niche.update(name: niche_params[:name], parent: parent)
+      redirect_to niche_url(@niche)
```
Comment (Sergyenko) [inline app/controllers/niches_controller.rb]: same as above

File: app/controllers/niches_controller.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  # DELETE /locations/1.json
+  def destroy
+    @niche.destroy
+    redirect_to niches_url
+  end
+
+  private
+
+  # Use callbacks to share common setup or constraints between actions.
+  def set_niche
```
Comment (Sergyenko) [inline app/controllers/niches_controller.rb:55]: Think about the creating helper for this code. As it's the same as here:
https://github.com/cybergizer/nicher/pull/28/files#diff-8cf35630782fed586aedb0f80d9fefbeR41

File: app/models/niche.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
+class Niche < ApplicationRecord
+  has_many :items, dependent: :delete_all
+  belongs_to :user
+  has_ancestry
+  validates :name, presence: true
+
+  def add(niche)
+    niche.parent = self
+    niche.save
```
Comment (Sergyenko) [inline app/models/niche.rb:14]: What if save is unsuccessful?

File: app/models/niche.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  def remove(niche)
+    our_niche = self
+    niche = our_niche.descendants.find_by(id: niche.id)
+    niche.parent = nil
+    niche.save
```
Comment (Sergyenko) [inline app/models/niche.rb]: What if save is unsuccessful?

File: app/models/niche.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  def delete(niche)
+    our_niche = self
+    niche = our_niche.descendants.find_by(id: niche.id)
+    niche.destroy
```
Comment (Sergyenko) [inline app/models/niche.rb]: What if destroy is unsuccessful?

File: app/models/user.rb [hunk 10]
```diff
@@ -1,6 +1,6 @@
 class User < ApplicationRecord
   has_many :items, dependent: :destroy
-
+  has_many :niches, class_name: 'Niche', dependent: :delete_all
```
Comment (Sergyenko) [inline app/models/user.rb]: Do you still need to specify "class_name: 'Niche'"?

File: app/views/niches/show.html.slim [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+table.table
+  thead
+    tr
+      th
+        | Title
+      th
+        | Description
+      th[colspan="3"]
+  tbody
+    - @niche.items.each do |item|
```
Comment (Sergyenko) [inline app/views/niches/show.html.slim:18]: What if @niche.items returns nil?

File: app/views/niches/show.html.slim [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+  h1
+    | Children Niches
+  table.table
+    thead
+      tr
+        th
+          | Name
+        th[colspan="3"]
+    tbody
+      - @niche.children.each do |niche|
```
Comment (Sergyenko) [inline app/views/niches/show.html.slim:40]: What if @niche.children returns nil?

File: app/views/shared/_menu.html.slim [hunk 13]
```diff
@@ -8,6 +8,10 @@
       a href="/" 
         i.material-icons widgets
         span Items
+    li
+      a href="/niches"
```
Comment (Sergyenko) [inline app/views/shared/_menu.html.slim]: Use rails URL helpers

**Review by Sergyenko — CHANGES_REQUESTED** [review]

File: spec/controllers/niches_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+      {
+        name: 'asafd',
+        user_id: user.id
+      }
+    }
+
+    it "updates the requested location" do
+      niche = Niche.create! valid_attributes
+      put :update, params: {id: niche.to_param, niche: new_attributes}, session: valid_session
+      niche.reload
```
Comment (anatoliliotych) [inline spec/controllers/niches_controller_spec.rb]: check that new attributes were applied

**Review by anatoliliotych — CHANGES_REQUESTED** [review]

**Review by Sergyenko — DISMISSED** [review]

File: app/assets/stylesheets/login.scss [hunk 15]
```diff
@@ -21,6 +21,7 @@ body, html {
   }
 }
 
+
```
Comment (anatoliliotych) [inline app/assets/stylesheets/login.scss]: why it is here?

File: app/controllers/niches_controller.rb [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    @niches = current_user.niches.reject(&:parent)
+  end
+
+  # GET /locations/1
+  # GET /locations/1.json
+  def show; end
+
+  # GET /locations/new
+  def new
+    @niche = current_user.niches.create
```
Comment (anatoliliotych) [inline app/controllers/niches_controller.rb]: it's kind of dangerous to create objects on form opening, do you remove them if user doesn't save form?

File: app/models/niche.rb [hunk 17]
```diff
@@ -0,0 +1,26 @@
+class Niche < ApplicationRecord
+  PARAMS = %i[name parent_id].freeze
+
+  has_many :items, dependent: :delete_all
```
Comment (anatoliliotych) [inline app/models/niche.rb:4]: try to keep these sections as separate blocks via empty lines

File: app/models/niche.rb [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+  validates :name, presence: true
+
+  def add(niche)
+    niche.parent = self
+    niche.save
+  end
+
+  def remove(niche)
+    our_niche = self
+    niche = our_niche.descendants.find_by(id: niche.id)
```
Comment (anatoliliotych) [inline app/models/niche.rb:19]: find_by can return nil, so you'll call parent and save on nil.

File: app/models/niche.rb [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+    niche.parent = nil
+    niche.save
+  end
+
+  def delete(niche)
+    our_niche = self
+    niche = our_niche.descendants.find_by(id: niche.id)
```
Comment (anatoliliotych) [inline app/models/niche.rb]: I'd remove this method, it's kind of strange to remove one niche through another.

**Review by anatoliliotych — COMMENTED** [review]
app/views/.DS_Store - should be ignored

File: spec/controllers/niches_controller_spec.rb [hunk 20]
```diff
@@ -0,0 +1,172 @@
+require 'rails_helper'
+
+RSpec.describe NichesController, type: :controller do
+  login_user
+
+  let(:user) { subject.current_user }
+  let(:niche) { create(:niche, user: user) }
+  let(:valid_attributes) { attributes_for(:niche).merge(user_id: user.id) }
+  let(:invalid_attributes) { attributes_for(:invalid_niche).merge(user_id: 1) }
```
Comment (anatoliliotych) [inline spec/controllers/niches_controller_spec.rb]: this is wrong way to make invalid params, what if user_id=1 will exist?

File: spec/controllers/niches_controller_spec.rb [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+
+    it 'assigns the niche' do
+      expect(assigns(:niche)).to be_a_new(Niche)
+    end
+
+    it "renders the #new view" do
+      expect(response).to render_template :new
+    end
+
+    it "redirects to new view page" do
```
Comment (anatoliliotych) [inline spec/controllers/niches_controller_spec.rb]: invalid example name! should be about valid response

File: spec/controllers/niches_controller_spec.rb [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+  describe 'GET #edit' do
+    before do
+      get :edit, params: { id: niche }
+    end
+
+    it 'assigns the niche' do
+      expect(assigns(:niche)).to eq(niche)
+    end
+
+    it "renders the @edit view" do
```
Comment (anatoliliotych) [inline spec/controllers/niches_controller_spec.rb]: @edit?

File: spec/factories/user.rb [hunk 23]
```diff
@@ -1,5 +1,5 @@
 FactoryBot.define do
-  factory :user do
+  factory :user, aliases: %i[author] do
```
Comment (anatoliliotych) [inline spec/factories/user.rb]: what does it mean author?

**Review by anatoliliotych — CHANGES_REQUESTED** [review]
why this file is removed `config/database.yml.example`?

**Comment by anatoliliotych** [comment]
lgtm-ship-it

**Review by anatoliliotych — APPROVED** [review]

**Pull request merged**